### PR TITLE
core: use multiple fee-recipient addresses

### PR DESCRIPTION
**Pull request by dB2510**
Uses multiple fee-recipient addresses from lock introduced in v1.5.

category: feature
ticket: #1651

### CODE REVIEW

File: app/app.go [hunk 1]
```diff
@@ -317,14 +317,21 @@ func wireCoreWorkflow(ctx context.Context, life *lifecycle.Manager, conf Config,
...
 	var (
-		corePubkeys       []core.PubKey
-		eth2Pubkeys       []eth2p0.BLSPubKey
-		pubshares         []eth2p0.BLSPubKey
-		allPubSharesByKey = make(map[core.PubKey]map[int]*bls_sig.PublicKey) // map[pubkey]map[shareIdx]pubshare
+		corePubkeys                  []core.PubKey
+		eth2Pubkeys                  []eth2p0.BLSPubKey
+		pubshares                    []eth2p0.BLSPubKey
+		allPubSharesByKey            = make(map[core.PubKey]map[int]*bls_sig.PublicKey) // map[pubkey]map[shareIdx]pubshare
+		feeRecipientAddsByEth2Pubkey = make(map[eth2p0.BLSPubKey]string)
```
Comment (xenowits) [inline app/app.go:331]:
```suggestion
		feeRecipientAddrsByEth2Pubkey = make(map[eth2p0.BLSPubKey]string)
```
similarly for `feeRecipientAddsByEth2Pubkey`

File: app/app.go [hunk 2]
```diff
@@ -772,7 +775,7 @@ func wireTracing(life *lifecycle.Manager, conf Config) error {
 
 // setFeeRecipient returns a slot subscriber for scheduler which calls prepare_beacon_proposer endpoint at start of each epoch.
 // TODO(dhruv): move this somewhere else once more use-cases like this becomes clear.
-func setFeeRecipient(eth2Cl eth2wrap.Client, pubkeys []eth2p0.BLSPubKey, feeRecipient string) func(ctx context.Context, slot core.Slot) error {
+func setFeeRecipient(eth2Cl eth2wrap.Client, pubkeys []eth2p0.BLSPubKey, feeRecipientByPubkey map[eth2p0.BLSPubKey]string) func(ctx context.Context, slot core.Slot) error {
```
Comment (corverroos) [inline app/app.go:778]: you are passing in pubkeys twice, only pass it in once.

File: core/validatorapi/teku.go [hunk 3]
```diff
@@ -43,10 +43,15 @@ type TekuProposerConfigProvider interface {
...
 
 func (c Component) TekuProposerConfig(ctx context.Context) (TekuProposerConfigResponse, error) {
+	var defaultFeeRecipientAddr string
+	for _, addr := range c.feeRecipientAddressByPubkey {
+		defaultFeeRecipientAddr = addr
+		break
+	}
 	resp := TekuProposerConfigResponse{
 		Proposers: make(map[string]TekuProposerConfig),
 		Default: TekuProposerConfig{ // Default doesn't make sense, disable for now.
```
Comment (corverroos) [inline core/validatorapi/teku.go:53]: as the comment says, "Default doesn't make sense, disable for now", rather just disable and use zero values for the rest.

File: core/validatorapi/validatorapi.go [hunk 4]
```diff
@@ -58,7 +58,7 @@ func NewComponentInsecure(_ *testing.T, eth2Cl eth2wrap.Client, shareIdx int) (*
 //
 //nolint:gocognit
 func NewComponent(eth2Cl eth2wrap.Client, allPubSharesByKey map[core.PubKey]map[int]*bls_sig.PublicKey,
-	shareIdx int, feeRecipientAddress string, builderAPI bool, seenPubkeys func(core.PubKey),
+	shareIdx int, feeRecipientAddressByPubkey map[core.PubKey]string, builderAPI bool, seenPubkeys func(core.PubKey),
```
Comment (corverroos) [inline core/validatorapi/validatorapi.go:61]: same here, rather not pass in pubkeys twice, wouldn't it be better to use a single map with both shares and feerecipients? The other option is pass in a callback for `feeRecipientFunc func(core.Pubkey)string`

Comment (dB2510) [reply core/validatorapi/validatorapi.go]: Thanks for pointing out, went with `feeRecipientFunc func(core.Pubkey)string`

File: core/validatorapi/teku.go [hunk 5]
```diff
@@ -36,7 +36,10 @@ type TekuBuilder struct {
 	Overrides map[string]string `json:"registration_overrides,omitempty"`
 }
 
-const gasLimit = 30000000
+const (
+	gasLimit    = 30000000
+	zeroAddress = "0x0000000000000000000000000000000000000000"
```
Comment (corverroos) [inline core/validatorapi/teku.go:41]: zero values mean golang zero values, so do not populate the `Default`field. Just make sure Teku validation doesn't fail in this case.

Comment (dB2510) [reply core/validatorapi/teku.go:41]: yup, tried that earlier unfortunately teku doesn't like that:
Screenshot 2023-01-25 at 4 23 41 PM

**Review by corverroos — APPROVED** [review]

**Pull request merged**